### ISSUE TITLE
BugSigDB 1.0 release

### ISSUE THREAD

**Issue by lgeistlinger** [waldronlab/BugSigDB#80]
Hi @jwokaty:

@lwaldron and I had started to discuss a release scheme for BugSigDB.

One idea was to follow Bioconductor's semi-annual release scheme, and have a stable release of BugSigDB signatures every half a year. We also discussed [zenodo](https://zenodo.org/) as the platform for hosting the stable release (= csv files for studies, experiments, and signatures).

A stable release is supposed to contain *all reviewed* content from BugSigDB up to a defined freeze date. For the BugSigDB 1.0 release this could encompass reviewed content up to the present date for simplicity, or if we wanted to synchronize with Bioconductor, up to the past 3.13 release date.

Would you like to go ahead and export the content, filter by date and review status, upload to zenodo, and include the stable release link under https://bugsigdb.org/Help:Export ?

Thanks!

**Comment by lgeistlinger** [comment]
Note, functionality from our [bugsigdbr](https://github.com/waldronlab/bugsigdbr) package will likely be helpful here, including `bugsigdbr::importBugSigDB` for pulling the data frame that can then be filtered along the usual lines, `bugsigdbr::getSignatures` to extract the signatures of the filtered df, and `bugsigdbr::writeGMT` to write a GMT file containing the signatures.

**Comment by lwaldron** [comment]
I added a new organization secret `ZENODO_DEPOSIT` with an access token that should allow depositing through the Zenodo from GitHub Actions in this repo, bugsigdbr, and bugphyzz. Zenodo also has an option for pulling releases directly from any release of a GitHub repo, although that would require either depositing a software + data repo, or creating separate data-only repos. Would be great if you can look into these options a bit @jwokaty then we can discuss the best way to do periodic data releases.

**Comment by lgeistlinger** [comment]
Hi @jwokaty @lwaldron : I thought I quickly check in with you as to what's the status of this and whether input/help from my side is required here?

**Comment by jwokaty** [comment]
@lgeistlinger I'd like to better understand what the release looks like. Will it be 3 csv files: one each for studies, experiments, and signatures? Or will there be different versions in content or file formats? 

There's also a place where I can add metadata. I'll try to write it using the website and the repository, but maybe I need some help with the following:

Are there any specific keywords should we associate with the data? This isn't required, but may help users find it.

* Is there a license associated with this data? Zenodo lists the following licenses (but I think we can do something else):
Creative Commons Attribution 4.0 International
Creative Commons Attribution 1.0 Generic
Creative Commons Attribution 2.0 Generic
Creative Commons Attribution 3.0 Unported
Creative Commons Attribution 3.0 Austria
Creative Commons Attribution 3.0 United States

**Comment by lgeistlinger** [comment]
Thanks for checking in @jwokaty.

> Will it be 3 csv files: one each for studies, experiments, and signatures? 

That's the core, yes. This is also what I think what `bugsigdbr` will pull at one point from zenodo. My thought would be that

```
bugsigdbr::importBugSigDB
```

will get a `version` argument so once the 1.0 release is out on zenodo, users could pull the stable release via:

```
bugsigdbr::importBugSigDB(version = "1.0")
```

Currently it directly pulls the bloody edge including unreviewed content from https://bugsigdb.org/Help:Export, which would then become available via:

```
bugsigdbr::importBugSigDB(version = "devel")
```

When preparing the three csv files for the 1.0 release, it's important to restrict to reviewed contents only. Let me know whether you have questions or whether it's easier if I provide these files.

In addition, I think we want to supplement the core release with GMT files containing the actual signatures.
There are many ways to extract signatures based on taxonomic considerations and `bugsigdbr::getSignatures` implements a bunch of those options. For non R-users, we however want to provide at least 3 GMT files I think:

- signatures containing mixed taxonomic levels,
- signatures harmonized on genus level
- (signatures harmonized on species level)

@lwaldron let us know if you have thoughts on this.

> Are there any specific keywords should we associate with the data? This isn't required, but may help users find it.

A couple that come to mind:

- human microbiome
- differential abundance
- gene set enrichment analysis

> Is there a license associated with this data? Zenodo lists the following licenses (but I think we can do something else):

@lwaldron might have opinions. I'd be good with Creative Commons Attribution 4.0 International.

**Comment by lwaldron** [comment]
@lgeistlinger would you write a simple .R script that dumps all the required files? I guess it'll be a few: gmt, cab, ncbi ID, names, genus, species, mixed. No need to put versioning or dates in file names, but maybe a comment line in the first line with the date, license, and reference to bugsigdb.org?

**Comment by lgeistlinger** [comment]
Sure.

**Comment by lgeistlinger** [comment]
That is basically done and part of bugsigdbr now:

https://github.com/waldronlab/bugsigdbr/blob/main/inst/scripts/dump_release.R

Call the script via: `Rscript dump_release.R <version> <output.directory>`

which will produce the following output files:

```
full_dump.tab 
bugsigdb_signatures_mixed_metaphlan.gmt
bugsigdb_signatures_mixed_ncbi.gmt
bugsigdb_signatures_mixed_taxname.gmt
bugsigdb_signatures_genus_metaphlan.gmt		
bugsigdb_signatures_genus_metaphlan_exact.gmt	
bugsigdb_signatures_genus_ncbi.gmt		
bugsigdb_signatures_genus_ncbi_exact.gmt	
bugsigdb_signatures_genus_taxname.gmt		
bugsigdb_signatures_genus_taxname_exact.gmt
bugsigdb_signatures_species_metaphlan.gmt
bugsigdb_signatures_species_metaphlan_exact.gmt
bugsigdb_signatures_species_ncbi.gmt
bugsigdb_signatures_species_ncbi_exact.gmt
bugsigdb_signatures_species_taxname.gmt
bugsigdb_signatures_species_taxname_exact.gmt
```

Pending #92 a filter by review status will be incorporated.

**Comment by jwokaty** [comment]
@lgeistlinger I am working on this at https://github.com/jwokaty/BugSigDBExports, which I will transfer to waldronlab when it's in a good state. I've created a GitHub action that we can run manually to generate exports with `dump_release.R`.

@lwaldron and I had discussed setting up the action to do a daily export that would be committed to BugSigDBExports. We would do a manual release to get into Zenodo.

However, the version number that is passed to `dump_release` is just a label, right? I thought if possible it might be better to associate the files with a date corresponding to reviewed content rather than a version number. When we do a release to Zenodo, there will still be a version number, but I think we want to be able to reproduce the files whether we get it from BugSigDBExports, Zenodo, or bugsigdbr.

**Comment by lgeistlinger** [comment]
I think that is great. 

> However, the version number that is passed to dump_release is just a label, right?

This is correct. The only place where the `version` argument is used in `dump_release.R` is the header line of the output files, eg [here](https://github.com/jwokaty/BugSigDBExports/blob/main/bugsigdb_signatures_genus_metaphlan.gmt):

```
# BugSigDB 0.0.1, License: Creative Commons Attribution 4.0 International, URL: https://bugsigdb.org
```

If you call the script with a slight notational abuse and provide a date instead of a version number to `dump_release.R`, the header line will accordingly change to eg:

```
# BugSigDB 2021-07-27, License: Creative Commons Attribution 4.0 International, URL: https://bugsigdb.org
```

and the argument can then also be renamed to `date` in the script.

**Comment by lwaldron** [comment]
It looks great! Nice use of the `.zenodo.json` file too.

**Comment by lwaldron** [comment]
And BTW I think the Zenodo dumps can just have a date within too, releases will have a version number from the tag but the files contain dates as usual.

**Comment by jwokaty** [comment]
We are still waiting on #92 to close before we publish because this may change the content, correct? I've transferred the BugSigDBExports to waldronlab and have scheduled it to do weekly exports on Sunday using dates. The first will be this Sunday. Does this issue really belong to the new repository? I can transfer it.

**Comment by lwaldron** [comment]
Yes, this issue does belong in https://github.com/waldronlab/BugSigDBExports.
